xarray/core/dataset.py

```diff
@@ -6629,7 +6629,7 @@ def interpolate_na(
             | None
         ) = None,
         **kwargs: Any,
-    ) -> Self:
+    ) -> Dataset:
```

Comment:
We generally prefer `Self` over the concrete type, since it allows subtyping. Any reason we're changing these?

Reply:
MyPy rejected return type "Self," I guess because this function is passing the self object through `missing._apply_over_vars_with_dim` which is not a method. (It also happens that `_apply_over_vars_with_dim` creates a copy.) I think it would be reasonable to make `_apply_over_vars_with_dim` a method, or maybe there is an existing method with the same functionality. I wasn't sure if it was a good idea to make code-affecting changes in a PR about typing.

Reply:
If possible, we should try and increase `Self` and fix the resulting errors with something like `T_Dataset`, or ignoring the resulting errors...

Reply:
I'm not sure I understand the first suggestion. However I can just add an ignore in cases like this.

Reply:
Sorry for being unclear!

I meant that `Self` is favored over `Dataset`, but they're not always compatible. When some of the code wants `Self` and some wants `Dataset`, we should favor `Self`. And then gradually `Self` becomes more widespread, though requires some `# type: ignore` along the transition...

(And `T_Dataset` is more compatible than with `Self` than `Dataset`; albeit a more complicated topic)

Does that make a bit more sense? The headline is "let's not replace `Self` with `Dataset`"...

Reply:
Got it! Thanks for the clarification :)